packages/firestore/lite/src/api/components.ts

```diff
@@ -18,32 +18,141 @@
+ * A concrete type describing all the values that can be applied via a
+ * user-supplied firestore.Settings object. This is a separate type so that
+ * defaults can be supplied and the value can be checked for equality.
+ */
+export class FirestoreSettings {
+  /** The hostname to connect to. */
+  readonly host: string;
+
+  /** Whether to use SSL when connecting. */
+  readonly ssl: boolean;
+
+  readonly cacheSizeBytes: number;
+
+  readonly experimentalForceLongPolling: boolean;
+
+  readonly experimentalAutoDetectLongPolling: boolean;
+
+  readonly ignoreUndefinedProperties: boolean;
+
+  // Can be a google-auth-library or gapi client.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  credentials?: any;
+
+  constructor(settings: PrivateSettings) {
+    if (settings.host === undefined) {
+      if (settings.ssl !== undefined) {
+        throw new FirestoreError(
+          Code.INVALID_ARGUMENT,
+          "Can't provide ssl option if host option is not set"
+        );
+      }
+      this.host = DEFAULT_HOST;
+      this.ssl = DEFAULT_SSL;
+    } else {
+      this.host = settings.host;
+      this.ssl = settings.ssl ?? DEFAULT_SSL;
+    }
+
+    this.credentials = settings.credentials;
+    this.ignoreUndefinedProperties = !!settings.ignoreUndefinedProperties;
+
+    if (settings.cacheSizeBytes === undefined) {
+      this.cacheSizeBytes = LRU_DEFAULT_CACHE_SIZE_BYTES;
+    } else {
+      if (
+        settings.cacheSizeBytes !== LRU_COLLECTION_DISABLED &&
+        settings.cacheSizeBytes < LRU_MINIMUM_CACHE_SIZE_BYTES
+      ) {
+        throw new FirestoreError(
+          Code.INVALID_ARGUMENT,
+          `cacheSizeBytes must be at least ${LRU_MINIMUM_CACHE_SIZE_BYTES}`
+        );
+      } else {
+        this.cacheSizeBytes = settings.cacheSizeBytes;
+      }
+    }
+
+    this.experimentalForceLongPolling = !!settings.experimentalForceLongPolling;
+    this.experimentalAutoDetectLongPolling = !!settings.experimentalAutoDetectLongPolling;
+
+    validateIsNotUsedTogether(
+      'experimentalForceLongPolling',
+      settings.experimentalForceLongPolling,
+      'experimentalAutoDetectLongPolling',
+      settings.experimentalAutoDetectLongPolling
+    );
+  }
+
+  isEqual(other: FirestoreSettings): boolean {
+    return (
+      this.host === other.host &&
+      this.ssl === other.ssl &&
+      this.credentials === other.credentials &&
+      this.cacheSizeBytes === other.cacheSizeBytes &&
+      this.experimentalForceLongPolling ===
+        other.experimentalForceLongPolling &&
+      this.experimentalAutoDetectLongPolling ===
+        other.experimentalAutoDetectLongPolling &&
+      this.ignoreUndefinedProperties === other.ignoreUndefinedProperties
+    );
+  }
+}
+
 // The components module manages the lifetime of dependencies of the Firestore
 // client. Dependencies can be lazily constructed and only one exists per
 // Firestore instance.
 
+export interface FirestoreService extends _FirebaseService {
```

Comment:
Comments, please. In particular what's a FirestoreService and how does it differ from a FirebaseFirestore instance?

Reply:
I added the following:

```
/**
 * An interface implemented by FirebaseFirestore that provides
 * compatibility with the usage in this file. 
 * 
 * This interface mainly exists to remove a cyclic dependency.
 */
```

I will explore doing the same FirebaseFirestore/FirebaseFirestoreImpl pattern in a follow-up PR.